libp2p/protocols/secure/noise.nim

```diff
@@ -0,0 +1,575 @@
+  when initiator:
+    hs.ss.mixKey(dh(hs.e.privateKey, hs.rs))
+  else:
+    hs.ss.mixKey(dh(hs.s.privateKey, hs.re))
+
+template dh_se: untyped =
+  trace "noise dh se"
+  # Calls MixKey(DH(s, re)) if initiator, MixKey(DH(e, rs)) if responder.
+  when initiator:
+    hs.ss.mixKey(dh(hs.s.privateKey, hs.re))
+  else:
+    hs.ss.mixKey(dh(hs.e.privateKey, hs.rs))
+
+template dh_ss: untyped =
+  trace "noise dh ss"
+  # Calls MixKey(DH(s, rs)).
+  hs.ss.mixKey(dh(hs.s.privateKey, hs.rs))
+
+template read_e: untyped =
+  trace "noise read e", size = msg.len
+
+  if msg.len < Curve25519Key.len:
+    raise newException(NoiseHandshakeError, "Noise E, expected more data")
+
+  # Sets re (which must be empty) to the next DHLEN bytes from the message. Calls MixHash(re.public_key).
+  copyMem(addr hs.re[0], addr msg[0], Curve25519Key.len)
+  msg = msg[Curve25519Key.len..msg.high]
+  hs.ss.mixHash(hs.re)
+
+template read_s: untyped =
+  trace "noise read s", size = msg.len
+  # Sets temp to the next DHLEN + 16 bytes of the message if HasKey() == True, or to the next DHLEN bytes otherwise.
+  # Sets rs (which must be empty) to DecryptAndHash(temp).
+  var temp: seq[byte]
+  if hs.ss.cs.hasKey:
+    if msg.len < Curve25519Key.len + ChaChaPolyTag.len:
+      raise newException(NoiseHandshakeError, "Noise S, expected more data")
+   
+    temp.setLen(Curve25519Key.len + 16)
+    copyMem(addr temp[0], addr msg[0], Curve25519Key.len + 16)
+    msg = msg[Curve25519Key.len + 16..msg.high]
+  else:
+    if msg.len < Curve25519Key.len:
+      raise newException(NoiseHandshakeError, "Noise S, expected more data")
+
+    temp.setLen(Curve25519Key.len)
+    copyMem(addr temp[0], addr msg[0], Curve25519Key.len)
+    msg = msg[Curve25519Key.len..msg.high]
+  var plain = hs.ss.decryptAndHash(temp)
+  copyMem(addr hs.rs[0], addr plain[0], Curve25519Key.len)
+
+proc receiveHSMessage(sconn: Connection): Future[seq[byte]] {.async.} =
+  var
+    besize: array[2, byte]
+    size: uint16
+  await sconn.readExactly(addr besize[0], 2)
+  bigEndian16(addr size, addr besize[0])
+  trace "receiveHSMessage", size
+  return await sconn.read(size.int)
+
+proc sendHSMessage(sconn: Connection; buf: seq[byte]) {.async.} =
+  var
+    lesize = buf.len.uint16
+    size: array[2, byte]
+  trace "sendHSMessage", size = lesize
+  bigEndian16(addr size[0], addr lesize)
+  await sconn.write(addr size[0], 2)
+  await sconn.write(buf)
+
+proc packNoisePayload(payload: openarray[byte]): seq[byte] {.inline.} =
+  if payload.len > uint16.high.int:
+    raise newException(NoiseOversizedPayloadError, "Trying to send an unsupported oversized payload over Noise")
+
+  let
+    noiselen = rand(2..31)
+
+  var
+    plen = payload.len.uint16
+    besize: array[2, byte]
+    noise = newSeq[byte](noiselen)
+
+  if randomBytes(noise) != noiselen:
+    raise newException(NoiseHandshakeError, "Failed to generate randomBytes")
+
+  bigEndian16(addr besize[0], addr plen)
```

Comment:
again, see `endians2` which would allow  somthing like `let besize = plen.toBytesBE(...)` which is less code and safer thanks to fewer mutations and pointer ops